### PR TITLE
IS31FL3741 Framebuffer Driver

### PR DESCRIPTION
**Pull request by gamblor21**
Similar to the RGBMatrix (and also borrowing from I2CDisplay) this is a native driver for the IS31FL3741 chipset. I am releasing this is draft for others to take a look at but still requires some cleanup and I have some questions about the best way to implement some options. I have only tested this on the Adafruit Eyeglasses as they are the only IS31 board I have.

The main question I still have outstanding is how to pass in mapping from the display to actual LEDs. The glasses matrix is currently in the code but I am thinking having it as a parameter to pass in may make more sense and support future uses.

Right now this is only set to compile in the [LED glasses driver](https://www.adafruit.com/product/5217) board.

```python
import is31fl3741
import displayio
import framebufferio
import board
import busio

displayio.release_displays()

i2c = busio.I2C(board.SCL, board.SDA, frequency=1000000)
is31 = is31fl3741.is31fl3741(width=54, height=15, i2c=i2c, scale=True, gamma=True)
display = framebufferio.FramebufferDisplay(is31, auto_refresh=True)

bitmap = displayio.Bitmap(display.width, display.height, 2)
palette = displayio.Palette(2)
palette[0] = 0x000000
palette[1] = 0x990099
tile_grid = displayio.TileGrid(bitmap, pixel_shader=palette)

#draw a line at the bottom of the display
for y in range (12,15):
    for x in range(54):
        bitmap[x,y] = 1

group = displayio.Group()
group.append(tile_grid)
display.show(group)
```

### CODE REVIEW

**Comment by gamblor21** [comment]
One open question to anyone who has a thought: Any idea how to allow someone to take advantage of the glasses' rings. You could use the python library to otherwise address them or maybe another helper python library that is aware of the displayio display (and which pixels overlap) would be beneficial?

File: shared-module/displayio/Palette.c [hunk 1]
```diff
@@ -89,6 +89,9 @@ bool displayio_palette_get_color(displayio_palette_t *self, const _displayio_col
     } else if (colorspace->grayscale) {
         size_t bitmask = (1 << colorspace->depth) - 1;
         *color = (self->colors[palette_index].luma >> colorspace->grayscale_bit) & bitmask;
+    } else if (colorspace->depth == 32) {
```
Comment (gamblor21) [inline shared-module/displayio/Palette.c:92]: This was set to 32 bits (as opposed to 24) as in other place some division took place and 24 would require more changes. Handling 24 bit color in 32 bits is done in other systems but not sure if it is the best option here or not?

File: shared-module/is31fl3741/is31fl3741.h [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    busio_i2c_obj_t *i2c;
+    busio_i2c_obj_t inline_i2c;
+    uint8_t device_address;
+    uint8_t bit_depth;
+    bool paused;
+    bool scale;
+    bool auto_gamma;
+} is31fl3741_is31fl3741_obj_t;
+
+static const uint16_t glassesmatrix_ledmap[18 * 5 * 3] = {
```
Comment (gamblor21) [inline shared-module/is31fl3741/is31fl3741.h:48]: Looking at ideas on how to pass this data into the display driver

File: shared-module/is31fl3741/is31fl3741.h [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+    285,   52,    286,   // (16,4)
+    65535, 65535, 65535, // (17,0) (clipped, corner)
+    270,   272,   271,   // (17,1) / left ring pixel 4
+    27,    29,    28,    // (17,2) / 5
+    23,    25,    24,    // (17,3) / 6
+    276,   22,    277,   // (17,4) / 7
+};
+
+// Gamma correction table
+static const uint8_t IS31GammaTable[256] = {
```
Comment (gamblor21) [inline shared-module/is31fl3741/is31fl3741.h:142]: Would it make sense to calculate this at run time (using RAM) vs compile time? A bonus of run-time is the gamma level could be user controller.

File: shared-module/is31fl3741/is31fl3741.h (hunk 2 above)
Comment (tannewt) [reply shared-module/is31fl3741/is31fl3741.h:48]: Passing in a byte buffer is how I'd do it. I'd assume we'd provide libraries that do the init. So, most folks wouldn't need to worry about it.

File: shared-module/is31fl3741/is31fl3741.h (hunk 3 above)
Comment (tannewt) [reply shared-module/is31fl3741/is31fl3741.h:142]: If we want user control then RAM is a good choice. I'd start simpler and just hard code it like you did here. The generation code will also take space so the savings will be < 256 bytes. Simplicity is worth whatever the difference is.

File: shared-bindings/is31fl3741/__init__.c [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+#include "py/runtime.h"
+
+#include "shared-bindings/is31fl3741/is31fl3741.h"
+
+//| """Low-level routines for bitbanged LED matrices"""
+//|
+
+STATIC const mp_rom_map_elem_t is31fl3741_module_globals_table[] = {
+    { MP_ROM_QSTR(MP_QSTR___name__), MP_ROM_QSTR(MP_QSTR_is31fl3741) },
+    { MP_ROM_QSTR(MP_QSTR_is31fl3741), MP_ROM_PTR(&is31fl3741_is31fl3741_type) },
```
Comment (tannewt) [inline shared-bindings/is31fl3741/__init__.c:39]: I think we usually capitalize letters.

```suggestion
    { MP_ROM_QSTR(MP_QSTR_IS31FL3741), MP_ROM_PTR(&is31fl3741_is31fl3741_type) },
```

File: shared-module/displayio/Palette.c (hunk 1 above)
Comment (tannewt) [reply shared-module/displayio/Palette.c:92]: I think I added this as 32 bits for the raspberry pi too. In that case, the last byte is actually alpha. So, this works well for me :-).

File: shared-module/is31fl3741/is31fl3741.c [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+                    bsum = bsum / 9;
+                    uint32_t color = 0;
+                    if (self->auto_gamma) {
+                        color = (IS31GammaTable[rsum] << 16) +
+                            (IS31GammaTable[gsum] << 8) +
+                            IS31GammaTable[bsum];
+                    } else {
+                        color = (rsum << 16) + (gsum << 8) + bsum;
+                    }
+                    is31fl3741_draw_pixel(self->i2c, self->device_address, x, y, color);
```
Comment (tannewt) [inline shared-module/is31fl3741/is31fl3741.c:183]: How is the speed of this? Would you consider buffering the whole memory and then blasting it all out at once?

**Comment by PaintYourDragon** [comment]
Trying this out. Is the scrolly font available?

**Comment by gamblor21** [comment]
> Trying this out. Is the scrolly font available?
[tfont.zip](https://github.com/adafruit/circuitpython/files/7567775/tfont.zip)

This was converted from your Arduino font (possibly badly) but seems to work. Let me know if the zip attachment of it worked if not I'll upload it elsewhere.

**Comment by gamblor21** [comment]
> Trying this out. Is the scrolly font available?

This is a quick copy of the test code I used (its messy)

```python
import is31fl3741
import displayio
import framebufferio
import board
import busio
import time
from adafruit_display_text import label
from adafruit_bitmap_font import bitmap_font
from rainbowio import colorwheel

displayio.release_displays()

i2c = busio.I2C(board.SCL, board.SDA, frequency=1000000)
is31 = is31fl3741.is31fl3741(width=54, height=15, i2c=i2c, scale=True, gamma=True)
display = framebufferio.FramebufferDisplay(is31, auto_refresh=True)

is31.brightness = 1.0

text = "HELLO FROM CIRCUITPYTHON ON NATIVE DISPLAYIO"
font = bitmap_font.load_font("tfont.bdf")
color = 0xBB00BB
text_area = label.Label(font, text=text, color=color)
x = 54
text_area.x = x
text_area.y = 8

group = displayio.Group()
group.append(text_area)
display.show(group)

width = text_area.bounding_box[2]
idx = 0
while True:
    #text_area.color = colorwheel(idx)
    idx += 1
    x = x - 1
    text_area.x = x
    if x == -width:
        x = 54
    time.sleep(0.01)
```

**Comment by gamblor21** [comment]
Hmm rebasing to try to fix possible conflicts I seemed to have added a lot to this PR... I'll see if I can figure out how to fix that or may have to ask (or just re-PR it I think it is ready to come out of draft)

**Comment by gamblor21** [comment]
[supporting.zip](https://github.com/adafruit/circuitpython/files/7577036/supporting.zip)

LED mapping python file and font file to help with anyone who wants to test.

Code above should work but the new initializer for the glasses is:

```python
from map import glassesmatrix_ledmap
is31 = is31fl3741.IS31FL3741(width=54, height=15, i2c=i2c, scale=True, gamma=True, mapping=glassesmatrix_ledmap)
```

**Comment by gamblor21** [comment]
Note: I will try to add in something for the ring lights but figured it was better to put this PR in now and expand on it. Will also have to decide how/where a helper library to initialize and include the mapping for glasses (and other boards?) exists.

File: shared-bindings/is31fl3741/__init__.c (hunk 4 above)
Comment (gamblor21) [reply shared-bindings/is31fl3741/__init__.c]: Fixed it.

File: shared-module/is31fl3741/is31fl3741.c (hunk 5 above)
Comment (gamblor21) [reply shared-module/is31fl3741/is31fl3741.c]: The speed seems to match the Arduino code. I could play with blasting it all out at once but haven't had a chance to.

File: ports/nrf/boards/microbit_v2/mpconfigboard.mk [hunk 6]
```diff
@@ -13,9 +13,11 @@ CIRCUITPY_BUILTINS_POW3=0
 CIRCUITPY_BUSDEVICE = 0
 CIRCUITPY_COUNTIO = 0
 CIRCUITPY_DISPLAYIO = 0
+CIRCUITPY_IS31FL3741 = 0
```
Comment (tannewt) [inline ports/nrf/boards/microbit_v2/mpconfigboard.mk:16]:
```suggestion
```

File: shared-bindings/is31fl3741/is31fl3741.c [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+#include "shared-bindings/is31fl3741/is31fl3741.h"
+// #include "shared-bindings/microcontroller/Pin.h"
+// #include "shared-bindings/microcontroller/__init__.h"
+#include "shared-bindings/util.h"
+#include "shared-module/displayio/__init__.h"
+#include "shared-module/framebufferio/__init__.h"
+#include "shared-module/framebufferio/FramebufferDisplay.h"
+#include "shared-bindings/busio/I2C.h"
+
+//| class is31fl3741:
```
Comment (tannewt) [inline shared-bindings/is31fl3741/is31fl3741.c:41]: Please search for all capitalization variations of this. They should all be the same (all caps) for the class, lower for the module.

File: shared-bindings/is31fl3741/is31fl3741.c [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+ */
+
+#include "py/obj.h"
+#include "py/objproperty.h"
+#include "py/runtime.h"
+#include "py/objarray.h"
+
+#include "shared-bindings/is31fl3741/is31fl3741.h"
+// #include "shared-bindings/microcontroller/Pin.h"
+// #include "shared-bindings/microcontroller/__init__.h"
```
Comment (tannewt) [inline shared-bindings/is31fl3741/is31fl3741.c:34]:
```suggestion
```

File: shared-module/is31fl3741/is31fl3741.h (hunk 2 above)
Comment (tannewt) [reply shared-module/is31fl3741/is31fl3741.h:50]: Remove this please since it is passed in now.

**Review by tannewt — CHANGES_REQUESTED** [review]
Just a few more things. Thanks for working on this!

File: shared-bindings/is31fl3741/is31fl3741.c (hunk 7 above)
Comment (gamblor21) [reply shared-bindings/is31fl3741/is31fl3741.c]: Think I caught them all now.

**Review by tannewt — APPROVED** [review]
Thank you for the updates! This is cool!

**Pull request merged**